File: packages/core-ts/src/encode.ts
// Semantic key encoders per SPEC.md §3.5 / §3.6 / appendix A.1.

import { blake3 } from "@noble/hashes/blake3.js";
import { randomHv, bind, type Hypervector } from "./hypervector.js";

const STRING_DOMAIN = "str:";
const LEVEL_DOMAIN = "lvl:";
const LEVELS = 100;

export function encodeString(s: string, dimension: number): Hypervector {
  const seed = blake3(new TextEncoder().encode(STRING_DOMAIN + s));
  return randomHv(seed, dimension);
}

// Binary thermometer + random projection embedding -> hypervector.
// See SPEC.md appendix A.1.
export function encodeEmbedding(embedding: readonly number[], dimension: number): Hypervector {
  let acc: Hypervector = new Uint8Array(dimension);
  const enc = new TextEncoder();
  for (let i = 0; i < embedding.length; i++) {
    const raw = embedding[i]!;
    const clamped = Math.max(-1, Math.min(1, raw));
    const level = Math.round(((clamped + 1) * (LEVELS - 1)) / 2);
    const seedString = `${LEVEL_DOMAIN}${i}:${level}`;
    const seed = blake3(enc.encode(seedString));
    const hv = randomHv(seed, dimension);
    acc = bind(acc, hv);
  }
  return acc;
}

export function isHypervector(x: unknown, dimension?: number): x is Hypervector {
  if (!(x instanceof Uint8Array)) return false;
  if (dimension !== undefined && x.length !== dimension) return false;
  return true;
}

File: packages/core-rs/src/encode.rs
//! Semantic key encoders. Per SPEC.md §3.5 / §3.6 / appendix A.1.

use crate::hypervector::{bind, random_hv, Hypervector};

const STRING_DOMAIN: &str = "str:";
const LEVEL_DOMAIN: &str = "lvl:";
const LEVELS: usize = 100;

/// Map a UTF-8 string to a hypervector via BLAKE3-derived randomness.
pub fn encode_string(s: &str, dim: usize) -> Hypervector {
    let mut seed = Vec::with_capacity(STRING_DOMAIN.len() + s.len());
    seed.extend_from_slice(STRING_DOMAIN.as_bytes());
    seed.extend_from_slice(s.as_bytes());
    let digest = blake3::hash(&seed);
    random_hv(digest.as_bytes(), dim)
}

/// Binary thermometer + random projection: map a `[-1, 1]` embedding to a
/// hypervector that preserves cosine similarity approximately.
///
/// See SPEC.md appendix A.1.
pub fn encode_embedding(embedding: &[f32], dim: usize) -> Hypervector {
    let mut acc: Hypervector = vec![0u8; dim];
    for (i, &raw) in embedding.iter().enumerate() {
        let clamped = raw.clamp(-1.0, 1.0);
        let level = (((clamped + 1.0) * ((LEVELS - 1) as f32) / 2.0).round() as i32)
            .clamp(0, (LEVELS - 1) as i32) as usize;
        let mut seed = Vec::with_capacity(LEVEL_DOMAIN.len() + 32);
        seed.extend_from_slice(LEVEL_DOMAIN.as_bytes());
        seed.extend_from_slice(i.to_string().as_bytes());
        seed.push(b':');
        seed.extend_from_slice(level.to_string().as_bytes());
        let digest = blake3::hash(&seed);
        let hv = random_hv(digest.as_bytes(), dim);
        acc = bind(&acc, &hv);
    }
    acc
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::hypervector::similarity;

    #[test]
    fn encode_string_is_deterministic() {
        let a = encode_string("hello", 4096);
        let b = encode_string("hello", 4096);
        let c = encode_string("world", 4096);
        assert_eq!(a, b);
        assert!(similarity(&a, &c) < 0.55);
    }

    #[test]
    fn embedding_encoding_round_trips_through_levels() {
        let v: Vec<f32> = (0..16).map(|i| (i as f32) / 16.0 - 0.5).collect();
        let a = encode_embedding(&v, 4096);
        let b = encode_embedding(&v, 4096);
        assert_eq!(a, b);
    }
}
